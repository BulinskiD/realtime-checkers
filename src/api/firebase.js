import app from "firebase/app";
import "firebase/storage";
import "firebase/auth";
import "firebase/firestore";

var firebaseConfig = {
  apiKey: "AIzaSyD3uWma_iug5o_0w6EmchIXp4qd054HqcE",
  authDomain: "realtime-checkers.firebaseapp.com",
  databaseURL: "https://realtime-checkers.firebaseio.com",
  projectId: "realtime-checkers",
  storageBucket: "realtime-checkers.appspot.com",
  messagingSenderId: "100295366319",
  appId: "1:100295366319:web:ec1c182d83b660e8"
};

// Initialize Firebase
app.initializeApp(firebaseConfig);

export const auth = app.auth();
export const firestore = app.firestore();
