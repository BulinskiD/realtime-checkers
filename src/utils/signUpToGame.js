import { firestore } from "../api/firebase";
import handleError from "./handleError";

export default async (email, gameID, players) => {
  const color = players.length === 0 ? "white" : "black";
  players.push({ email, started: false, color });
  try {
    await firestore
      .collection("games")
      .doc(gameID)
      .update({
        players: players
      });
    await firestore
      .collection("users")
      .doc(email)
      .update({
        gameID: gameID
      });
  } catch (error) {
    handleError(error);
  }
};
