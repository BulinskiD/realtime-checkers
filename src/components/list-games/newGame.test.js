import React from "react";
import { act } from "react-dom/test-utils";
import ConnectedNewGame, { NewGame } from "./newGame";
import configureMockStore from "redux-mock-store";
import signUpToGame from "../../utils/signUpToGame";
import { firestore } from "../../api/firebase";
import { shallow } from "enzyme";

jest.mock("../../utils/signUpToGame");
jest.mock("../../api/firebase");
firestore.collection = jest.fn();
const add = jest.fn();
firestore.collection.mockReturnValue({ add });

describe("NewGame", () => {
  it("should display Link returnToGame when gameID is set", () => {
    const component = shallow(<NewGame user={{ gameID: "sda" }} />);
    expect(component.find("Link").props().to).toBe("/game/sda");
    expect(component.find("Button").text()).toBe("Powróć do gry!");
  });

  it("should display newGameForm when gameID is null", () => {
    const component = shallow(<NewGame user={{ gameID: null }} />);
    expect(component).toMatchSnapshot();
  });

  it("should show whole form after click on button", () => {
    const component = shallow(<NewGame user={{ gameID: null }} />);
    act(() => {
      component
        .find("Button")
        .at(0)
        .simulate("click");
    });
    expect(component).toMatchSnapshot();
  });

  it("should add new game after sending form with correct data", async () => {
    const selectGame = jest.fn();
    const prom = Promise.resolve({ id: "testID" });
    const sign = Promise.resolve();
    signUpToGame.mockReturnValue(sign);
    add.mockReturnValue(prom);
    const history = { push: jest.fn() };

    const component = shallow(
      <NewGame
        user={{ gameID: null, email: "test" }}
        selectGame={selectGame}
        history={history}
      />
    );
    act(() => {
      component
        .find("Button")
        .at(0)
        .simulate("click");
    });

    act(() => {
      component
        .find(".title")
        .props()
        .onChange({ target: { value: "test" } });
    });
    const preventDefault = jest.fn();

    act(() => {
      component
        .find("Styled(Form)")
        .props()
        .onSubmit({ preventDefault });
    });

    expect(preventDefault).toHaveBeenCalledTimes(1);
    await prom;
    await sign;
    expect(selectGame).toHaveBeenCalledWith("testID");
    expect(signUpToGame).toHaveBeenCalledWith("test", "testID", []);
    expect(history.push).toHaveBeenCalledWith("/game/testID");
  });

  it("should reset input on error", async () => {
    const selectGame = jest.fn();
    const prom = Promise.resolve({ id: "testID" });
    const sign = Promise.reject();
    signUpToGame.mockReturnValue(sign);
    add.mockReturnValue(prom);
    const history = { push: jest.fn() };

    const component = shallow(
      <NewGame
        user={{ gameID: null, email: "test" }}
        selectGame={selectGame}
        history={history}
      />
    );
    act(() => {
      component
        .find("Button")
        .at(0)
        .simulate("click");
    });

    act(() => {
      component
        .find(".title")
        .props()
        .onChange({ target: { value: "test" } });
    });
    const preventDefault = jest.fn();

    act(() => {
      component
        .find("Styled(Form)")
        .props()
        .onSubmit({ preventDefault });
    });
    try {
      await prom;
      await sign;
      expect(selectGame).toHaveBeenCalledWith("testID");
    } catch (error) {
      expect(component.find(".title").text()).toBe("");
      expect(signUpToGame).toHaveBeenCalledWith("test", "testID", []);
      expect(history.push).toHaveBeenCalledTimes(0);
    }
  });
});

describe("MapStateToProps", () => {
  const mockedStore = configureMockStore();

  it("should map state to props correctly", () => {
    const store = mockedStore({ user: { gameID: "test" } });
    const component = shallow(<ConnectedNewGame store={store} />);
    expect(component.find("NewGame").props().user).toStrictEqual({
      gameID: "test"
    });
  });
});
