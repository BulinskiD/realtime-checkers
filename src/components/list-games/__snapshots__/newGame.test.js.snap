// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewGame should display newGameForm when gameID is null 1`] = `
<Fragment>
  <h4
    className="my-3"
  >
    Nowa gra:
    <Button
      active={false}
      className="float-right"
      disabled={false}
      onClick={[Function]}
      type="button"
      variant="dark"
    >
      +
    </Button>
  </h4>
  <hr />
  <Styled(Form)
    onSubmit={[Function]}
    show={0}
  >
    <FormLabel
      column={false}
      srOnly={false}
    >
      Nazwa
    </FormLabel>
    <FormControl
      aria-label="title"
      className="mb-4 title"
      onChange={[Function]}
      placeholder="Nazwa"
      type="text"
      value=""
    />
    <Button
      active={false}
      className="float-right"
      disabled={false}
      type="submit"
      variant="dark"
    >
      Dodaj!
    </Button>
  </Styled(Form)>
</Fragment>
`;

exports[`NewGame should show whole form after click on button 1`] = `
<Fragment>
  <h4
    className="my-3"
  >
    Nowa gra:
    <Button
      active={false}
      className="float-right"
      disabled={false}
      onClick={[Function]}
      type="button"
      variant="dark"
    >
      -
    </Button>
  </h4>
  <hr />
  <Styled(Form)
    onSubmit={[Function]}
    show={1}
  >
    <FormLabel
      column={false}
      srOnly={false}
    >
      Nazwa
    </FormLabel>
    <FormControl
      aria-label="title"
      className="mb-4 title"
      onChange={[Function]}
      placeholder="Nazwa"
      type="text"
      value=""
    />
    <Button
      active={false}
      className="float-right"
      disabled={false}
      type="submit"
      variant="dark"
    >
      Dodaj!
    </Button>
  </Styled(Form)>
</Fragment>
`;
